xarray/core/missing.py

```diff
@@ -407,13 +408,13 @@ def _floatize_x(x, new_x):
     x = list(x)
     new_x = list(new_x)
     for i in range(len(x)):
-        if x[i].dtype.kind in 'Mm':
+        if _contains_datetime_like_objects(x[i]):
             # Scipy casts coordinates to np.float64, which is not accurate
             # enough for datetime64 (uses 64bit integer).
             # We assume that the most of the bits are used to represent the
             # offset (min(x)) and the variation (x - min(x)) can be
             # represented by float.
-            xmin = np.min(x[i])
+            xmin = x[i].min()
             x[i] = to_numeric(x[i], offset=xmin, dtype=np.float64)
```

Comment:
What will happen if x and xnew are different data type?
(E.g. date time vs float)

Reply:
Good question -- it seems the following was indeed allowed before this PR:
```
In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: import xarray as xr

In [4]: times = pd.date_range('2000', periods=5)

In [5]: da = xr.DataArray(np.arange(5), coords=[times], dims=['time'])

In [6]: da
Out[6]:
<xarray.DataArray (time: 5)>
array([0, 1, 2, 3, 4])
Coordinates:
  * time     (time) datetime64[ns] 2000-01-01 2000-01-02 ... 2000-01-05

In [7]: da.interp(time=0.5)
Out[7]:
<xarray.DataArray ()>
array(5.787037037037037e-15)
Coordinates:
    time     float64 0.5
```
where `0.5` is implicitly interpreted as nanoseconds since the first date in the original coordinate.  

This was surprising to me at first (I'm used to thinking about float-representations of dates in the [CF-context](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#time-coordinate), where an array of floats representing time must be accompanied by a units attribute that explicitly defines how those floats are to be interpreted as dates).  If this use case is important to preserve I can make sure that it continues to work (and I can add tests to make sure we don't break it later on).

Reply:
This behavior is actually error prone.
I originally tried to design this method to work as similar to `sel`, so maybe we can learn from that method also for this edge case.

Reply:
It's fine to convert between int and float, but we should avoid automatically converting datetime units. That case should raise an error.

Reply:
Sounds good -- I wasn't sure if I was missing something. I'll make sure this raises an informative `TypeError`.

Reply:
With 5377b16, the result from my example now looks like:
```
In [6]: da.interp(time=0.5)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-9dfef9ee2abb> in <module>()
----> 1 da.interp(time=0.5)

/Users/spencerclark/xarray-dev/xarray/xarray/core/dataarray.pyc in interp(self, coords, method, assume_sorted, kwargs, **coords_kwargs)
   1021         ds = self._to_temp_dataset().interp(
   1022             coords, method=method, kwargs=kwargs, assume_sorted=assume_sorted,
-> 1023             **coords_kwargs)
   1024         return self._from_temp_dataset(ds)
   1025

/Users/spencerclark/xarray-dev/xarray/xarray/core/dataset.pyc in interp(self, coords, method, assume_sorted, kwargs, **coords_kwargs)
   2005                     var_indexers = {k: _validate_interp_indexer(
   2006                         maybe_variable(obj, k), v) for k, v
-> 2007                                     in indexers.items() if k in var.dims}
   2008                     variables[name] = missing.interp(
   2009                         var, var_indexers, method, **kwargs)

/Users/spencerclark/xarray-dev/xarray/xarray/core/dataset.pyc in <dictcomp>((k, v))
   2005                     var_indexers = {k: _validate_interp_indexer(
   2006                         maybe_variable(obj, k), v) for k, v
-> 2007                                     in indexers.items() if k in var.dims}
   2008                     variables[name] = missing.interp(
   2009                         var, var_indexers, method, **kwargs)

/Users/spencerclark/xarray-dev/xarray/xarray/core/dataset.pyc in _validate_interp_indexer(x, new_x)
   1995                                'interpolate to must be either datetime '
   1996                                'strings or datetimes. '
-> 1997                                'Instead got\n{}'.format(new_x))
   1998             else:
   1999                 return (x, new_x)

TypeError: When interpolating over a datetime-like coordinate, the coordinates to interpolate to must be either datetime strings or datetimes. Instead got
<xarray.Variable ()>
array(0.5)
```
@fujiisoup does this look like the right place for this error message to you?

Reply:
Thanks, @spencerkclark. It looks very nice to me.